### ISSUE TITLE
Cannot inject JhiLanguageService into any core component

### ISSUE THREAD

**Issue by codecholeric** [jhipster/generator-jhipster#10476]
##### **Overview of the issue**

Regarding the Angular template: Cannot inject `JhiLanguageService` into any core component. As far as I see, as soon as `JhiLanguageService` is injected into any eagerly bootstrapped component, the app does not work anymore.

More precisely, within `AccountService.fetch()` `this.languageService` is undefined which causes any authentication to fail.

##### **Motivation for or Use Case**

What we try to achieve is, while loading the application, to set the language to the current browser language. Out attempt was to add the following to `JhiMainComponent`:

```
export class JhiMainComponent implements OnInit {
  constructor(
      ...
      private translateService: TranslateService,
      private languageService: JhiLanguageService) {}

  ngOnInit() {
    ...
    const languageKey = this.translateService.getBrowserLang();
    this.languageService.changeLanguage(languageKey);
  }
}
```

##### **Reproduce the error**

Create a new JHipster 6.3.1 application with the `angular` template. Then add the above lines to `JhiMainComponent`.

##### **Suggest a Fix**

As far as I see, the problem is that the constructor of `JhiLanguageService` calls `this.translateService.setDefaultLang(this.currentLang)` which in turn tries to load the language from the backend via HTTP. However due to the registered 

```
{
  provide: HTTP_INTERCEPTORS,
  useClass: AuthExpiredInterceptor,
  multi: true
}
```

that in turn demands 

```
constructor(private loginService: LoginService)
```

where `LoginService` demands

```
constructor(private accountService: AccountService, private authServerProvider: AuthServerProvider) {}
```

and `AccountService` in turn demands 

```
constructor(private languageService: JhiLanguageService, private sessionStorage: SessionStorageService, private http: HttpClient) {}
```

This causes a cyclic dependency

```
JhiLanguageService -> TranslateService -> AuthExpiredInterceptor -> LoginService -> AccountService -> JhiLanguageService
```

Thus the requested `JhiLanguageService` in the constructor of `AccountService` will be undefined and the whole authentication process breaks.

I don't know if our approach is wrong here (where to add our default behavior about languages), however the behavior and type of error seems something unwanted.

The only solutions I can think of is either

* switch the language later / lazily (where would be the best place to do this?)
* bypass the interceptors to load languages for internationalization (it does not seem useful to pass through an `AuthExpiredInterceptor` when loading languages that you need before the login has ever passed)

##### **JHipster Version(s)**

6.3.1

##### **JHipster configuration**

 jhipster info
INFO! Using JHipster version installed locally in current project's node_modules
INFO! Executing jhipster:info
INFO! Options: from-cli: true
Welcome to the JHipster Information Sub-Generator

##### **JHipster Version(s)**

```
app@0.0.1-SNAPSHOT /home/user/workspace/app
└── generator-jhipster@6.3.0 

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.mycompany.myapp",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "6.3.0",
    "applicationType": "monolith",
    "baseName": "app",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "no",
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "gradle",
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "...",
    "embeddableLaunchScript": false,
    "useSass": true,
    "clientPackageManager": "npm",
    "clientFramework": "angularX",
    "clientTheme": "none",
    "clientThemeVariant": "",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "entitySuffix": "",
    "dtoSuffix": "DTO",
    "otherModules": [],
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": ["en", "de"],
    "blueprints": []
  }
}

</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>

</pre>
</details>

##### **Environment and Tools**

openjdk version "11.0.4" 2019-07-16
OpenJDK Runtime Environment (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3)
OpenJDK 64-Bit Server VM (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3, mixed mode, sharing)

git version 2.17.1

node: v10.15.0

npm: 6.11.3

Docker version 17.12.0-ce, build c97c6d6

docker-compose version 1.16.1, build 6d1ac21

##### **Browsers and Operating System**

```
Version 76.0.3809.100 (Official Build) Built on Ubuntu , running on Ubuntu 18.04 (64-bit)
Ubuntu 18.04.3 LTS 
```

-   [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by antonioortizpola** [comment]
I do not know if this could help you, but I found a similar case some time ago working with JHipster

[Acces to Principal outside the 'router-outlet' in main component in JHipster Angular](https://stackoverflow.com/questions/53587152/acces-to-principal-outside-the-router-outlet-in-main-component-in-jhipster-ang)

> I can inject the Principal in any component rendered inside the router-outlet, however, if i try to inject it in the main component, an error is shown whenever i try to login/logout (if i refresh the page everything shows fine).

In the end I created another service:

- create a service with the variables
- inject this service in the main component
- edited the login service

**Comment by codecholeric** [comment]
Thanks for the suggestion! I've tried it, but at least injecting another service holding `JhiLanguageService` did not do the trick. I've noticed though, that the same cycle seems to be resolved with JHipster 6.2.0, so I somehow suggest it is something that used to be more "lazy" and now kicks in earlier (changes on `language-helper` removing the workaround with the observable maybe?)

However I've found one way to fix it: Make the resolution of `JhiLanguageService` within `AccountService` lazy, i.e. change the constructor

```
private lazyLanguageService;

constructor(private injector: Injector, private sessionStorage: SessionStorageService, private http: HttpClient) {
  this.lazyLanguageService = () => injector.get(JhiLanguageService)
}
```

and then dynamically resolve `JhiLanguageService` where it is needed:

```
if (this.userIdentity.langKey) {
  const langKey = this.sessionStorage.retrieve('locale') || this.userIdentity.langKey;
  this.lazyLanguageService().changeLanguage(langKey);
}
```

Anyway, I don't like this solution a lot, because it changes JHipster generated code, so we will have merge conflicts during the next update. Thus I'd still be happy about a clean solution!

**Comment by vishal423** [comment]
> bypass the interceptors to load languages for internationalization (it does not seem useful to pass through an AuthExpiredInterceptor when loading languages that you need before the login has ever passed)

I am not clear on this as you should be able to load languages resource bundle as those are not protected by login and associated interceptors. You can follow same by manually changing language in the drop down (before login). Initial language bundle chosen on the application load is specified by `defaultI18nLang: 'en'` in core module. In case, you want to change to browser default, then, you need to change here.

**Comment by kaidohallik** [comment]
@vishal423 submitted a good answer.

Just FYI I add comment about my investigation results.

Seems that this comes from Angular HTTP interceptors DI problem.

I debugged HTTP interceptors part and found out that:
* in the current code all HTTP interceptors are constructed twice
* after adding `JhiLanguageService` to `MainComponent` all HTTP interceptors are constructed once

I debugged `AuthExpiredInterceptor` and related services DI and found out that:
* `loginService` in `AuthExpiredInterceptor`:
    * in the current code: undefined in the first execution of the constructor, correct value in the second execution of the constructor, Angular issue which deals with a similar topic: https://github.com/angular/angular/issues/25590
    * after using `JhiLanguageService` in the `MainComponent`, `loginService` has correct value in the only execution of the constructor
* `LoginService` is always constructed once and `accountService` has correct value both in the current code and after using `JhiLanguageService` in the `MainComponent`
* `AccountService` is always constructed once and `languageService` in the `AccountService` constructor:
    * in the current code is available in the constructor
    * after using `JhiLanguageService` in the `MainComponent` is undefined in the constructor. Seems that Angular is able to cure HTTP interceptors dependencies if they are undefined (by recreating interceptors) but not nested services dependencies.

More than 1 workaround exist to that Angular HTTP interceptors DI problem. Below 2 possible workarounds.

Workaround 1: add `JhiLanguageService` to app module `providers` array before HTTP interceptors, after that `JhiLanguageService` is always available in HTTP interceptors DI.

Workaroud 2: as `JhiLanguageService` is not needed in `AccountService` constructor then you can get it from `Injector` just before using that (as @codecholeric described in the previous comment), but I believe that more optimized code for that is:
```
if (this.languageService === undefined) {
  this.languageService = this.injector.get(JhiLanguageService);
}
```

**Comment by codecholeric** [comment]
@vishal423 thanks for the answer and the tips. I wasn't exactly sure, how I would apply your suggestion about `defaultI18nLang`. We don't want to "force" a default language on the user, but take the user's browser language as default for our application.
The config containing `defaultI18nLang` goes into `NgJhipsterModule` very early (module import of `CoreModule`), and since I need to inject `TranslateService` to get the browser language, it seems impossible to manipulate the language there.

What I tried now, and what seems to work, is to inject `JhiConfigService` into the constructor of `AppModule` and then changing `configService.getConfig().defaultI18nLang`. Is this how you would imagine it? I couldn't find any config object I could directly inject, since `JhiModuleConfig` is not the same object that is used by the `JhiConfigService`.

@kaidohallik thanks for your analysis effort! To me the behavior seems to be somewhat strange. Even though resource bundles should not be protected, as @vishal423 pointed out, it seems evident that the interceptors are instantiated and injected when the `TranslateService` makes its GET request.
I agree that simply setting the `languageService`, if it is undefined, is more efficient. However, I found when you have to add this guard to multiple places, it affects maintainability, so I would always rather use something like a "memoizing supplier" (callback that fetches the value only during the first call). Since the class is pretty small and the check only needs to be performed in one place, it doesn't make much difference though :wink:

**Comment by kaidohallik** [comment]
I was able to make HTTP interceptors DI process stable after I removed `init()` from `JhiLanguageService` constructor https://github.com/jhipster/ng-jhipster/blob/master/src/language/language.service.ts#L31 and put that into generated app constructor.  
@codecholeric thanks for providing useful information for that enhancement.  
Providing PR-s to `ng-jhipster` and `generator-jhipster` (this needs new release of the `ng-jhipster`) to fix this issue.
